chain/index/api.go

```diff
@@ -0,0 +1,347 @@
+	}
+
+	executionTs, err := si.getNextTipset(ctx, ts)
+	if err != nil {
+		return xerrors.Errorf("failed to get next tipset for height %d: %w", ts.Height(), err)
+	}
+
+	// given that `ts` is on the canonical chain and `executionTs` is the next tipset in the chain
+	// `ts` can not have reverted events
+	var hasRevertedEventsInTipset bool
+	err = si.stmts.hasRevertedEventsInTipsetStmt.QueryRowContext(ctx, tsKeyCid.Bytes()).Scan(&hasRevertedEventsInTipset)
+	if err != nil {
+		return xerrors.Errorf("failed to check if there are reverted events in tipset for height %d: %w", ts.Height(), err)
+	}
+	if hasRevertedEventsInTipset {
+		return xerrors.Errorf("index corruption: reverted events found for an executed tipset %s at height %d", tsKeyCid, ts.Height())
+	}
+
+	executedMsgs, err := si.executedMessagesLoaderFunc(ctx, si.cs, ts, executionTs)
+	if err != nil {
+		return xerrors.Errorf("failed to load executed messages for height %d: %w", ts.Height(), err)
+	}
+
+	var (
+		totalEventsCount       = uint64(0)
+		totalEventEntriesCount = uint64(0)
+	)
+	for _, emsg := range executedMsgs {
+		totalEventsCount += uint64(len(emsg.evs))
+		for _, ev := range emsg.evs {
+			totalEventEntriesCount += uint64(len(ev.Entries))
+		}
+	}
+
+	if totalEventsCount != indexedData.nonRevertedEventCount {
+		return xerrors.Errorf("event count mismatch for height %d: chainstore has %d, index has %d", ts.Height(), totalEventsCount, indexedData.nonRevertedEventCount)
+	}
+
+	totalExecutedMsgCount := uint64(len(executedMsgs))
+	if totalExecutedMsgCount != indexedData.nonRevertedMessageCount {
+		return xerrors.Errorf("message count mismatch for height %d: chainstore has %d, index has %d", ts.Height(), totalExecutedMsgCount, indexedData.nonRevertedMessageCount)
+	}
+
+	if indexedData.nonRevertedEventEntriesCount != totalEventEntriesCount {
+		return xerrors.Errorf("event entries count mismatch for height %d: chainstore has %d, index has %d", ts.Height(), totalEventEntriesCount, indexedData.nonRevertedEventEntriesCount)
+	}
+
+	return nil
+}
+
+func (si *SqliteIndexer) backfillMissingTipset(ctx context.Context, ts *types.TipSet) (*types.IndexValidation, error) {
+	// backfill the tipset in the Index
+	parentTs, err := si.cs.GetTipSetFromKey(ctx, ts.Parents())
+	if err != nil {
+		return nil, xerrors.Errorf("failed to get parent tipset at height %d: %w", ts.Height(), err)
+	}
+
+	executionTs, err := si.getNextTipset(ctx, ts)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to get next tipset at height %d: %w", ts.Height(), err)
+	}
+
+	backfillFunc := func() error {
+		return withTx(ctx, si.db, func(tx *sql.Tx) error {
+			if err := si.indexTipsetWithParentEvents(ctx, tx, ts, executionTs); err != nil {
+				return xerrors.Errorf("error indexing (ts, executionTs): %w", err)
+			}
+
+			if err := si.indexTipsetWithParentEvents(ctx, tx, parentTs, ts); err != nil {
+				return xerrors.Errorf("error indexing (parentTs, ts): %w", err)
+			}
+
+			return nil
+		})
+	}
```

Comment:
@aarshkshah1992 I'm using this now to backfill from a snapshot, along with https://github.com/filecoin-project/lotus/pull/12614, and the logging is weird:

```
2024-10-18 15:50:23.706 ✓ Epoch 4362804 ({"TipSetKey":[...],"Height":4362804,"IndexedMessagesCount":95,"IndexedEventsCount":195,"IndexedEventEntriesCount":1070,"Backfilled":true,"IsNullRound":false})
2024-10-18 15:50:23.709 ✓ Epoch 4362803 ({"TipSetKey":[...],"Height":4362803,"IndexedMessagesCount":98,"IndexedEventsCount":205,"IndexedEventEntriesCount":1113,"Backfilled":false,"IsNullRound":false})
```

i.e. every second one is `"Backfilled":false`. Because it has to recompute state to do it from snapshot, it's taking a few seconds to do each tipset, because it has to do two in one go, then I get the second one following up really quickly because it's already been done.

This is not great UX and will definitely lead to confusion. It's also going to result in one more tipset recomputed than you asked for, _or_, you'll get an error reported about the wrong tipset if you backfill right up until the last tipset you have (i.e. it'll do `indexTipsetWithParentEvents` for `executionTs` fine, then fail on `indexTipsetWithParentEvents` for `parentTs` but that failure will be reported for the tipset you asked to backfill which won't have actually failed.

Do we have to do the second `indexTipsetWithParentEvents` here?

Reply:
@rvagg 

I mean deferred execution means that the events for T are available at T+1. So, when we're indexing T+1 -> it does make sense to index events for T and to index events for T -> we gotta index T.

We can improve UX here by not indexing T at all when T+1 is being indexed but not backfilling something that logically makes sense just to improve a log message also dosent sound great.

To use your favourite emoji -> 🤷 -> if you have strong opinions here -> I am happy to go along with it.

Reply:
I think we should view this through the same lens as the query APIs - when you ask eth_getLogs or GetActorEventsRaw you ask for a specific epoch and get that epoch's events. I think we should do the same here. If we backfill for an epoch, then it should backfill for the events that you would get if you queried that epoch for its events.

1. Backfill only 4000000
2. Query 4000000 for events and get them
3. Query 4000001 or 40000002 for events and I shouldn't get them because I didn't backfill those

Something like that.

Reply:
Umm..yeah makes sense. No strong objections here. Will make the change on Monday. Let's keep this comment unresolved to track.

Reply:
This is done. We do not backfill the parent tipset of the requested epoch anymore.